keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,109 @@
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, with the introduction of Jobs and Cronjobs it became clear that they're not handled very well. If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a non-sidecar container in almost all respects apart from the fact that when all non-sidecar containers have permanently terminated all running sidecar containers will also be terminated with a reason of `Completed`. The sidecars will be terminated by receiving a SIGTERM and will obey the `terminationGracePeriodSeconds` defined on the pod before being sent a SIGKILL. Sidecar containers will be terminated in the same way as non-sidecars when a pod is deleted.
```

Comment:
What is the run sequence between sidecar container and init container? Should sidecar container run first and terminate last to simplify the entire logic? Or leave init container the first to run, then sidecar container, then app containers? Or at the start time, treat sidecar container the same as the regular ones?

Reply:
So currently it would go init containers start -> init containers finish -> sidecar and normal containers start -> normal containers finish -> sidecars finish. 
I think it's nice if init always run first. that means you can have sidecars that rely on an init container completing successfully.
There's a separate discussion to be had if init sidecars should be a thing.

Reply:
Ideally sidecars would run after all the init containers have finished. They should be treated the same as "normal" application containers except for not blocking Pod termination.